pkg/addonfactory/template_api_agentaddon.go

```diff
@@ -0,0 +1,465 @@
+package addonfactory
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/valyala/fasttemplate"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	addonv1alpha1client "open-cluster-management.io/api/client/addon/clientset/versioned"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+
+	"open-cluster-management.io/addon-framework/pkg/agent"
+	"open-cluster-management.io/addon-framework/pkg/utils"
+)
+
+const (
+	NodePlacementPrivateValueKey = "__NODE_PLACEMENT"
+	RegistriesPrivateValueKey    = "__REGISTRIES"
+)
+
+// templateBuiltinValues includes the built-in values for crd template agentAddon.
+// the values for template config should begin with an uppercase letter, so we need
+// to convert it to Values by JsonStructToValues.
+// the built-in values can not be overridden by getValuesFuncs
+type templateCRDBuiltinValues struct {
+	ClusterName           string `json:"CLUSTER_NAME,omitempty"`
+	AddonInstallNamespace string `json:"INSTALL_NAMESPACE,omitempty"`
+}
+
+// templateDefaultValues includes the default values for crd template agentAddon.
+// the values for template config should begin with an uppercase letter, so we need
+// to convert it to Values by JsonStructToValues.
+// the default values can be overridden by getValuesFuncs
+type templateCRDDefaultValues struct {
+	HubKubeConfigPath     string `json:"HUB_KUBECONFIG,omitempty"`
+	ManagedKubeConfigPath string `json:"MANAGED_KUBECONFIG,omitempty"`
+}
+
+type CRDTemplateAgentAddon struct {
+	getValuesFuncs     []GetValuesFunc
+	trimCRDDescription bool
+
+	hubKubeClient kubernetes.Interface
+	addonClient   addonv1alpha1client.Interface
+	addonName     string
+	agentName     string
+}
+
+// NewCRDTemplateAgentAddon creates a CRDTemplateAgentAddon instance
+func NewCRDTemplateAgentAddon(
+	addonName string,
+	hubKubeClient kubernetes.Interface,
+	addonClient addonv1alpha1client.Interface,
+	getValuesFuncs ...GetValuesFunc,
+) *CRDTemplateAgentAddon {
+	a := &CRDTemplateAgentAddon{
+		getValuesFuncs:     getValuesFuncs,
+		trimCRDDescription: true,
+
+		hubKubeClient: hubKubeClient,
+		addonClient:   addonClient,
+		addonName:     addonName,
+		agentName:     utilrand.String(5),
+	}
+
+	return a
+}
+
+func (a *CRDTemplateAgentAddon) Manifests(
+	cluster *clusterv1.ManagedCluster,
+	addon *addonapiv1alpha1.ManagedClusterAddOn) ([]runtime.Object, error) {
+
+	template, err := utils.GetDesiredAddOnTemplate(a.addonClient, addon)
```

Comment:
it is better to use lister here. you could define a getter interface, and start informer in the manager.